thinc/backends/ops.py

```diff
@@ -500,6 +568,18 @@ def alloc(
         else:
             return self.xp.empty(shape, dtype=dtype)
 
+    def reshape1(self, array: ArrayXd, d0: int) -> Array1d:
+        return cast(Array1d, self.reshape(array, (d0,)))
```

Comment:
It's a bit of a shame that the type system cannot express something like: if we pass in a `Floats3d`, then we return a `Floats1d`. Maybe it's something that we could explore in the future. E.g. Rust's `ndarray` library makes the dimensionality a type parameter. Very simplified and more in Python-ese:

```Array[Type, Dim]```

So, `Floats2d` would be

```
type Floats2d = Array[float32, Ix2]
```

then a method like this one could be rewritten like:

```
def reshape1(self, array: Array[T, Dim], d0: int) -> Array[T, Ix1]:
```

Definitely out-of-scope for this PR, but perhaps something that is worth considering in the future if MyPy supports it.

Reply:
That's what you'd use the `overrides` for though right?

Reply:
That is my understanding of what the `overrides` do. In this example, however, the method name is actually specifying how the array should be reshaped, which is outside the domain of mypy.

Reply:
> That's what you'd use the `overrides` for though right?

On every method, you'll have to enumerate all possible types of T. Overrides look useful when you really have different types, but unwieldy when you have different parametrizations of the same generic types.

> In this example, however, the method name is actually specifying how the array should be reshaped, which is outside the domain of mypy.

Maybe I made the example more complicated by giving the ideal scenario :). A simpler variant would be:

```
def reshape1(self, array: ArrayXd[T], d0: int) -> Array1d[T]:
```

I think this is also the approach numpy is taking. See e.g.:

https://github.com/numpy/numpy/pull/17719

They also want to add the shape to the type, like my earlier example (`Array[T, Dim`), but they are waiting on a PEP for that.